flag/flag.go

```diff
@@ -139,118 +159,176 @@ func (s *Set) parse() error {
+			}
+		case isTextM:
+			{
+				// Make sure our newVal value actually points to something.
+				newValPtr.Elem().Set(reflect.New(ft.Elem()))
+				newVal := newValPtr.Elem().Interface()
+				s.Flags.Var(marshalWrapper{v: newVal.(encoding.TextUnmarshaler)}, name, help)
+				continue
 			}
-			s.Flags.Int(name, fvToInt(), help)
+		case fieldVal.Type() == timeDuration:
+			s.Flags.Duration(name, fieldVal.Interface().(time.Duration), help)
+			continue
+		default:
 		}
-	case reflect.Int64:
-		s.Flags.Int64(name, fieldVal.Int(), help)
-	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32:
-		{
-			fvToInt := func() uint {
-				switch v := fieldVal.Interface().(type) {
-				case uint:
-					return v
-				case uint8:
-					return uint(v)
-				case uint16:
-					return uint(v)
-				case uint32:
-					return uint(v)
-				default:
-					return 0
+
+		k := ft.Kind()
+		for k == reflect.Ptr {
+			ft = ft.Elem()
+			k = ft.Kind()
+		}
+
+		switch k {
+		case reflect.String:
+			s.Flags.String(name, fieldVal.Interface().(string), help)
+		case reflect.Bool:
+			s.Flags.Bool(name, fieldVal.Interface().(bool), help)
+		case reflect.Float64:
+			s.Flags.Float64(name, fieldVal.Interface().(float64), help)
+		case reflect.Float32:
+			s.Flags.Float64(name, float64(fieldVal.Interface().(float32)), help)
+		case reflect.Complex64:
+			s.Flags.Var(&complex64Var{c: fieldVal.Interface().(complex64)}, name, help)
+		case reflect.Complex128:
+			s.Flags.Var(&complex128Var{c: fieldVal.Interface().(complex128)}, name, help)
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32:
+			{
+				fvToInt := func() int {
+					switch v := fieldVal.Interface().(type) {
+					case int:
+						return v
+					case int8:
+						return int(v)
+					case int16:
+						return int(v)
+					case int32:
+						return int(v)
+					default:
+						return 0
+					}
 				}
+				s.Flags.Int(name, fvToInt(), help)
+			}
+		case reflect.Int64:
+			s.Flags.Int64(name, fieldVal.Int(), help)
+		case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32:
+			{
+				fvToInt := func() uint {
+					switch v := fieldVal.Interface().(type) {
+					case uint:
+						return v
+					case uint8:
+						return uint(v)
+					case uint16:
+						return uint(v)
+					case uint32:
+						return uint(v)
+					default:
+						return 0
+					}
+				}
+				s.Flags.Uint(name, fvToInt(), help)
+			}
+		case reflect.Uint64:
+			s.Flags.Uint64(name, fieldVal.Interface().(uint64), help)
+		case reflect.Slice, reflect.Map:
+			switch ft {
+			case stringSlice:
+				s.Flags.Var(&stringSliceFlag{ptr.(*[]string)}, name, help)
+			case mapStringStringSlice:
+				s.Flags.Var(&mapStringStringSliceFlag{ptr.(*map[string][]string)}, name, help)
+			case mapStringString:
+				s.Flags.Var(&mapStringStringFlag{ptr.(*map[string]string)}, name, help)
+			case stringSet:
+				s.Flags.Var(&stringSetFlag{ptr.(*map[string]struct{})}, name, help)
+			default:
+				return fmt.Errorf("unhandled type %s",
```

Comment:
nit: Let's make this `fmt.Errorf` call one line.

Reply:
Good catch. Done